### ISSUE TITLE
FO - Missing 404 for category page that does not exist

### ISSUE THREAD

**Issue by Robin-Fischer-PS** [PrestaShop/PrestaShop#24145]
#### Describe the bug

On the FO, when trying to access a category that does not exist, an exception are displayed instead of 404 page (exception in debug mode, browser 404 page without debug mode).

#### Expected behavior

When accessing a non existent category page on FO we should have a in-shop 404 page and not an exception (or a brower 404 page).

#### Steps to Reproduce

1. Enable debug mode
2. In FO, try to access a category that does not exist URL (for example : [URL]/fr/18-test) : The following exception is displayed : 

![image](https://user-images.githubusercontent.com/55688534/115556798-55f59100-a2b1-11eb-8f26-8ff31a388dbf.png)

On 1.7.7 we are redirected to the 404 page : [URL]/fr/index.php?controller=404

**Screenshots**

See above

#### Additional information

* PrestaShop version: 1.7.8
* PHP version: N/A

**Comment by Robin-Fischer-PS** [comment]
Similar to #24143 but not the same error :)

**Comment by matks** [comment]
Probably due to a edge-case introduced in https://github.com/PrestaShop/PrestaShop/pull/21094/

What happens:
- URL matches a category structure so front CategoryController is called
- CategoryController in `init()`finds there is no category so it calls `$this->setTemplate('errors/404');`
- then `CategoryController::getTemplateVarPage()` is called, and unfortunately it calls `getCanonicalURL()` which attempts to load the `$category` object however it's not valid

The logic in `Link::getCategoryLink()` considers "if object is not loaded correctly it must be an array" (but it's not)

Possible solution:
```
// CategoryController.php

    public function getCanonicalURL()
    {
        if ($this->category->id === null) {
            return '';
        }
```

**Comment by Progi1984** [comment]
Fixed by @atomiix in #24270
